src/generate.rs

```diff
@@ -356,16 +355,71 @@ fn register_block(registers: &[Register], defs: &Defaults) -> Result<Tokens> {
     let mut i = 0;
     // offset from the base address, in bytes
     let mut offset = 0;
-    for register in util::expand(registers) {
-        let pad = if let Some(pad) = register.offset.checked_sub(offset) {
+    let mut registers_expanded = vec![];
+
+    for register in registers {
+        match *register {
+            Register::Single(ref _into) => registers_expanded.push(register.clone()),
+            Register::Array(ref _into, ref array_info) => {
+                let index_convertible = if let Some(ref indexes) = array_info.dim_index {
+                    let mut index_iter = indexes.iter();
+                    let mut previous = 0;
+
+                    let get_number_option = |element_option: Option<&String>| -> Option<usize> {element_option
+                                                                        .and_then(|element| match element.parse::<usize>() {
+                                                                            Ok(i) => Some(i),
+                                                                            _ => None,
+                                                                        })
+                    };
+
+                    let mut index_convertible_temp = match get_number_option(index_iter.next()) {
+                        Some(0) => true,
+                        _ => false,
+                    };
+                    
+                    for element in index_iter {
+                        if !index_convertible_temp { break; }
+
+                        index_convertible_temp = match get_number_option(Some(element)) {
```

Comment:
this `get_number_option(Some(element))` would be simpler as `element.parse().ok()`